### ISSUE TITLE
Which interactive mapping library or dashboard platform to use?

### ISSUE THREAD

**Issue by Robinlovelace** [cyipt/actdev#19]
Pros and cons of different options. Which to go for? Starter for 10 in roughly descending order of preference:

- I think mapbox's main library has been forked and that the open source community is rallying round this (not a JS developer but noticed some things on social media): https://github.com/maplibre/maplibre-gl-js
- Leaflet
- Cesium
- Any others?

Thoughts @mvl22? Patrick raised the question and thought it would be good to open it up.

**Comment by mvl22** [comment]
Depends what we want to achieve. We haven't defined a proposed UI yet. Technology should obvious follow that.

We have capability in Leaflet and Mapbox GL JS so I would strongly favour using one of those two.

NB Mapbox GL v1 is fine and will remain supported for a long while. It doesn't have to be used with Mapbox's own tiles.

**Comment by Robinlovelace** [comment]
> Depends what we want to achieve.

A fast-to-develop and easy-to-maintain/extend codebase for visualising datasets to change planning decisions so that they enable walking and cycling uptake and reduce car dependency associated with new developments, as set-out in #8. 

UI specific aims has been discussed in #11. In addition to visualising those layers (that I've just updated, heads-up @mvl22 ) and providing widgets (sliders, dropdown menus, check boxes etc), what else do we need to decide?

Interested in +s and - of each and how could the balance switch, i.e. what features could only one library deliver.

**Comment by Robinlovelace** [comment]
If we're going for more of a dashboard approach this looks like a good option with support for maps and DBs to avoid writing boilerplate code: https://superset.apache.org/

**Comment by Robinlovelace** [comment]
And of course there always shiny 🙂 would need to work out how to get embedded maps to consume data from a dynamic API.

**Comment by mem48** [comment]
As I understand it Mapbox GL v2 is still free to use just not open source. The only significant feature I can see in v2 is the 3D terrain. I can see that being useful as hilliness matters for cycling. @mvl22 are you considering 3D terrain for cyclestreets?

**Comment by mem48** [comment]
I've done a bit of reading on how Mapbox GL JS v2 implements 3D terrain, documenting here in case anybody is interested.

3D terrain is served as DEM of raster tiles and the rest of the layers are draped over the terrain, the problems for self-hosting are:

1. MapBox uses WEBP image format (like PNG but with better compression). Which is not widely supported in GIS software. There is R support https://cran.r-project.org/web/packages/webp/index.html

2. gdal2tiles (the usual way of making raster tiles) only supports 8-bit images (pixel value 0-255) so you can't have a DEM with any elevations above 255 m. Thus the top would be cut off all your mountains, and no support for areas below sea level.  A possible workaround would be to scale your DEM from 0-255, then use the scaling options in Mapbox GL JS to compensate. In the UK that would mean scaling 0-1,300m so you would have about 5m vertical resolution. I'd expect this would give hills obvious steps, and so would be a pretty poor solution.

I'm sure you could write custom code around these problems, but it would be quite a big task.

**Comment by Robinlovelace** [comment]
There's a solid example in deck.gl: https://deck.gl/examples/terrain-layer/

**Comment by Robinlovelace** [comment]
And check-out Cesium, I made this walk through of a walk we did in summer as a test: https://cesium.com/ion/stories/viewer/?id=644bd731-3ae6-41bc-8e92-3491e3b5f791

**Comment by mem48** [comment]
The deck.gl example requires a Mapbox API key so isn't self hosted, similarly for Cesium. If you have the funds to use mapbox then use them and avoid all the hassel of hosting your own server.

**Comment by Robinlovelace** [comment]
I've got a mapbox api key and have almost 50k free hits to go this month. We have funds if needs be:

![image](https://user-images.githubusercontent.com/1825120/102389590-3977de00-3fcb-11eb-8e49-66a771838460.png)


Details on generous free tier that should cover this project: https://www.mapbox.com/pricing/

**Comment by Robinlovelace** [comment]
If needs be. Don't think we need 3D or fancy maps at all for that matter and that we should think about the core functionality and 'MVP' tasks: https://github.com/cyipt/actdev/labels/mvp

On a related note (#17) what do you think of this fairly simple approach to visualising the JTS data @mem48 ?: https://apps.saferactive.org/actdev-shiny-demo/

At present keeping it simple and mature, which means Leaflet, is my thinking but happy to be persuaded to use other approaches.

**Comment by mem48** [comment]
In January I'm starting on my carbon mapping project I'm planning to use Mapbox and adapt my covid maps and the techniques @layik and I worked on super simple only needs a small amount of code see https://www.wisemover.co.uk/creds/covid/ happy to share code / experience

**Comment by mvl22** [comment]
To reiterate, I think the technology approach cannot be decided on until we know what the functionality will be. There's lots above I could comment on, but again I don't think it's appropriate to speculate on technologies before user needs are properly discussed and prototyped.

> visualising datasets to change planning decisions so that they enable walking and cycling uptake and reduce car dependency associated with new developments

This is a very helpful summary of the outcomes/impact, but it is not a definition of functionality.

> UI specific aims has been discussed in #11.

Ticket 11 refers to a prototype of just showing data on a map. That does not imply a good tool ready for widespread use.

For instance, in CyIPT we ended up with a layer-based website which makes it easy to access that data, but I would still argue it is quite 'prototypey' because it doesn't really aim to guide the user towards a set of clear research questions - it just presents the data and assumes the users will somehow work out what's important. It contains a whole pile of buttons, each of which has equal visual prominence, when in fact some controls are actually for more important than others. A good UI should in my view be trying actively to orientate a user.

For instance, UI questions that I expect will arise during the UI prototyping phase might be:

- What specific questions and angle should the user be able to ask? In other words, what actually is the functionality (as distinct from what data is available)?
- Do we want a map which is nationally focussed (Bikedata, StreetFocus, CycleStreets, etc. or regionally focussed (PCT)?
- Do we have a map that enables an initial orientation of the data but then allows people to move into a more specific interrogation system?
- Do we start with a set of buttons that effectively form an entry point to wizards to guide people to a view?
- Do we start a location search as a starting point (see FixMyStreet, StreetFocus, etc.) or just dump people directly into a UI?
- Do we start with a map but have particular layers significantly more prominent?
- Does the user actually click on a map for data to appear in a highly customised format, or is it pre-presented for browsing?
- Does the map contain a set of controls to be operated in parallel or is something more linear wanted?

These are all things we can't yet answer but will arise in the process of looking at what data we have, seeing what findings could be interrogable from it, and designing a user workflow.

If we just want to throw a load of data on a map, which is fine for merely interrogating the data in an alpha manner, I would probably go for Mapbox GL v1, and we already have experience in that, but I hope, having properly budgeted for UI development, we aren't ultimately planning to release a tool that is just a load of data thrown on a map :)

**Comment by Robinlovelace** [comment]
These are great questions that can be the basis of a detailed specification.

**Comment by mvl22** [comment]
> These are great questions that can be the basis of a detailed specification.

Agreed at today's meeting these will be discussed fully at the 29th January workshop meeting.

**Comment by Robinlovelace** [comment]
Closing as it seems this has been decided and this issue is now largely superseded by #37. Seems MapBox will be used. Sounds good to me.
